File: packages/engine/src/tools/email.ts
/**
 * `email.send` tool — send a transactional email via the configured mailer,
 * with per-org rate-limit gating + best-effort usage telemetry.
 *
 * Used by: `packages/engine/src/tool-registry.ts` (spreads `emailTools`).
 */

import { z } from "zod";
import { RATE_LIMIT_WINDOW_MS } from "../constants";
import { getMailer } from "../mailer";
import { getEngineRateLimiter } from "../rate-limit";
import { getEmailUsageRecorder } from "../email-usage";
import { defineTool, envPositiveInt, type ToolExecutionContext } from "./tool-types";

/**
 * Fires the registered email-usage recorder, swallowing any failure so
 * a recorder bug can't break the `email.send` tool path. Skipped when
 * no recorder is registered (unit tests).
 */
async function fireEmailRecorder(input: {
  orgId: string;
  executionContext: ToolExecutionContext;
  to: string;
  from: string;
  provider: "resend" | "sendgrid" | "simulator" | "noop";
  providerMessageId?: string;
  ok: boolean;
  error?: string;
  latencyMs: number;
}): Promise<void> {
  const recorder = getEmailUsageRecorder();
  if (!recorder) return;
  try {
    await recorder({
      orgId: input.orgId,
      runId: input.executionContext.runId,
      nodeId: input.executionContext.nodeId,
      workflowId: input.executionContext.workflowId,
      to: input.to,
      from: input.from,
      provider: input.provider,
      providerMessageId: input.providerMessageId,
      ok: input.ok,
      error: input.error,
      latencyMs: input.latencyMs,
    });
  } catch {
    // Telemetry must never break the tool. Drop silently.
  }
}

const emailSendInput = z.object({
  /** Recipient email address. v1 ships single-recipient only; multi-to lands when the AC asks. */
  to: z.string().min(1),
  /**
   * Sender email address. When omitted, the tool falls back to
   * `process.env.JANUSLY_MAILER_FROM` at execute time. Resend's
   * sandbox sender is `onboarding@resend.dev`; SendGrid requires a
   * DNS-verified domain.
   */
  from: z.string().min(1).optional(),
  /** Subject line. Capped at RFC 5322's 998-char line length. */
  subject: z.string().min(1).max(998),
  /** Plain-text body. Required when `html` is not supplied. */
  text: z.string().max(200_000).optional(),
  /** HTML body. Required when `text` is not supplied. */
  html: z.string().max(500_000).optional(),
  /**
   * Provider metadata: Resend `tags`, SendGrid `custom_args`. Kept
   * intentionally small because it is copied into outbound provider
   * payloads and billing telemetry.
   */
  metadata: z.record(z.string().min(1).max(64), z.string().max(256))
    .refine((value) => Object.keys(value).length <= 20, {
      message: "email.send metadata supports at most 20 entries",
    })
    .optional(),
}).refine(
  (input) => Boolean(input.text || input.html),
  { message: "email.send requires `text` or `html` (or both)." },
);

const emailSendOutput = z.object({
  /** True iff the provider accepted the email for delivery. */
  ok: z.boolean(),
  /** Which mailer handled the call. `"noop"` when no API key was configured. */
  provider: z.enum(["resend", "sendgrid", "simulator", "noop"]),
  /** Provider-assigned id; populated on success. */
  providerMessageId: z.string().optional(),
  /** Failure reason; populated when `ok === false`. Mirrors the AGENTS.md AI-fallback contract for write-side tools. */
  error: z.string().optional(),
});

export const emailTools = {
  "email.send": defineTool({
    name: "email.send",
    description: "Send a transactional email via the configured mailer (Resend or SendGrid).",
    inputSchema: emailSendInput,
    outputSchema: emailSendOutput,
    inputExample: { to: "user@example.com", subject: "Hello", text: "Body of the email." },
    writeSide: true,
    async execute(input, _context, executionContext) {
      const start = Date.now();
      const orgId = executionContext.orgId;
      const limiter = getEngineRateLimiter();
      const rateLimitPerMin = executionContext.email?.rateLimitPerMin
        ?? envPositiveInt("JANUSLY_EMAIL_RATE_LIMIT_PER_MIN", 100);

      // Per-org rate gate. The injected limiter throws when over
      // limit; the tool wrapper converts the throw to a clean
      // `{ ok: false, error }` envelope so the AI-fallback contract
      // holds and the workflow run doesn't fail.
      if (orgId && limiter) {
        try {
          await limiter("email.send", orgId, { windowMs: RATE_LIMIT_WINDOW_MS, max: rateLimitPerMin });
        } catch (err) {
          const error = err instanceof Error ? err.message : "Rate limit exceeded";
          await fireEmailRecorder({
            orgId,
            executionContext,
            to: input.to,
            from: input.from ?? executionContext.email?.from ?? process.env.JANUSLY_MAILER_FROM ?? "onboarding@resend.dev",
            provider: "noop",
            ok: false,
            error,
            latencyMs: Date.now() - start,
          });
          return { ok: false, provider: "noop" as const, error };
        }
      }

      const from = input.from ?? executionContext.email?.from ?? process.env.JANUSLY_MAILER_FROM ?? "onboarding@resend.dev";
      const mailer = getMailer(executionContext.email?.provider);
      const result = await mailer.send({
        to: input.to,
        from,
        subject: input.subject,
        text: input.text,
        html: input.html,
        metadata: input.metadata,
      });
      const latencyMs = Date.now() - start;

      // Best-effort audit row. Wrapped in try/catch inside the helper
      // so a recorder failure can't break the tool path. Skipped when
      // there's no orgId (unit tests, ad-hoc paths).
      if (orgId) {
        await fireEmailRecorder({
          orgId,
          executionContext,
          to: input.to,
          from,
          provider: result.provider,
          providerMessageId: result.ok ? result.providerMessageId : undefined,
          ok: result.ok,
          error: result.ok ? undefined : result.error,
          latencyMs,
        });
      }

      if (result.ok) {
        return { ok: true, provider: result.provider, providerMessageId: result.providerMessageId };
      }
      return { ok: false, provider: result.provider, error: result.error };
    },
  }),
};

File: internal/tools/email.go
// email.send — the transactional mailer tool implements the contract's
// tools/email.ts + mailer.ts. Four providers resolved AT CALL TIME:
// resend / sendgrid (real, both egressing through the guarded Post seam
// — no vendor SDK), simulator (explicit local-stack gate, never
// implicit), and noop (the safe default: {ok:false, "Mailer not
// configured"} — the write-side AI-fallback contract holds without a
// throw). Per-org rate gate (family "email", default 100/min) and
// best-effort usage telemetry ride the shared integration deps.
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"net/mail"
	"os"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"
)

// EmailSettings are the tenant-resolved posture knobs (org config
// email.provider / email.from with env fallbacks — resolved by the
// engine deps builder).
type EmailSettings struct {
	Provider string
	From     string
}

const (
	emailSubjectMax  = 998
	emailTextMax     = 200_000
	emailHTMLMax     = 500_000
	emailMetadataMax = 20
	emailAddressMax  = 320
)

var emailMetadataKey = regexp.MustCompile(`^[A-Za-z0-9_-]{1,64}$`)

func validEmailMailbox(value string) bool {
	if value == "" || value != strings.TrimSpace(value) || len(value) > emailAddressMax || crlfPattern.MatchString(value) {
		return false
	}
	address, err := mail.ParseAddress(value)
	return err == nil && address.Name == "" && address.Address == value
}

func validateEmailInput(input map[string]any, options InputValidationOptions) error {
	for _, field := range []string{"to", "from"} {
		raw, present := input[field]
		if !present || isDeferredWholeTemplate(raw, options) {
			continue
		}
		if !validEmailMailbox(raw.(string)) {
			return fmt.Errorf("%s must be one valid email address of at most %d bytes", field, emailAddressMax)
		}
	}
	if raw, present := input["subject"]; present && !isDeferredWholeTemplate(raw, options) {
		subject := raw.(string)
		if strings.TrimSpace(subject) == "" || len(subject) > emailSubjectMax || crlfPattern.MatchString(subject) {
			return fmt.Errorf("subject must be non-empty, single-line, and at most %d bytes", emailSubjectMax)
		}
	}

	textRaw, textPresent := input["text"]
	htmlRaw, htmlPresent := input["html"]
	textPotential := textPresent && isDeferredWholeTemplate(textRaw, options)
	htmlPotential := htmlPresent && isDeferredWholeTemplate(htmlRaw, options)
	if textPresent && !textPotential {
		text := textRaw.(string)
		if len(text) > emailTextMax {
			return fmt.Errorf("text exceeds %d bytes", emailTextMax)
		}
		textPotential = strings.TrimSpace(text) != ""
	}
	if htmlPresent && !htmlPotential {
		html := htmlRaw.(string)
		if len(html) > emailHTMLMax {
			return fmt.Errorf("html exceeds %d bytes", emailHTMLMax)
		}
		htmlPotential = strings.TrimSpace(html) != ""
	}
	if !textPotential && !htmlPotential && (options.RequireAll || textPresent || htmlPresent) {
		return fmt.Errorf("text or html is required")
	}

	if raw, present := input["metadata"]; present && !isDeferredWholeTemplate(raw, options) {
		metadata := raw.(map[string]any)
		if len(metadata) > emailMetadataMax {
			return fmt.Errorf("metadata supports at most %d entries", emailMetadataMax)
		}
		for key, rawValue := range metadata {
			value, ok := rawValue.(string)
			if !ok || !emailMetadataKey.MatchString(key) || value == "" || value != strings.TrimSpace(value) ||
				utf8.RuneCountInString(value) > 256 || crlfPattern.MatchString(value) {
				return fmt.Errorf("metadata must use safe 1..64 byte keys and trimmed single-line string values of at most 256 characters")
			}
		}
	}
	return nil
}

func emailTools() []Definition {
	unavailable := func(_ context.Context, _ map[string]any) (map[string]any, error) {
		return map[string]any{"ok": false, "provider": "noop", "error": "integration tools require run context"}, nil
	}
	return []Definition{{
		Name:        "email.send",
		Description: "Send a transactional email via the configured mailer (Resend or SendGrid).",
		Required:    []string{"to", "subject"},
		Optional:    []string{"from", "text", "html", "metadata"},
		Fields: []Field{
			{Name: "to", Type: "string", Required: true},
			{Name: "subject", Type: "string", Required: true},
			{Name: "from", Type: "string"},
			{Name: "text", Type: "string"},
			{Name: "html", Type: "string"},
			{Name: "metadata", Type: "object"},
		},
		InputExample: map[string]any{"to": "user@example.com", "subject": "Hello", "text": "Body of the email."},
		Validate:     validateEmailInput,
		WriteSide:    true,
		Execute:      unavailable,
	}}
}

// simulatorEndpoint mirrors the contract's explicit local-stack gate.
func simulatorEndpoint(path string) string {
	if !localIntegrationSimulatorEnabled() {
		return ""
	}
	base := strings.TrimSuffix(os.Getenv("JANUSLY_LOCAL_INTEGRATION_SIMULATOR_URL"), "/")
	if base == "" {
		return ""
	}
	return base + path
}

// resolveMailerProvider ports getMailer's resolution ladder.
func resolveMailerProvider(override string) string {
	requested := strings.ToLower(override)
	if requested == "" {
		requested = strings.ToLower(os.Getenv("JANUSLY_MAILER_PROVIDER"))
	}
	switch {
	case requested == "noop":
		return "noop"
	case requested == "resend" && os.Getenv("RESEND_API_KEY") != "":
		return "resend"
	case requested == "sendgrid" && os.Getenv("SENDGRID_API_KEY") != "":
		return "sendgrid"
	case requested == "simulator" && simulatorEndpoint("/email/send") != "":
		return "simulator"
	default:
		return "noop"
	}
}

func emailEnvelope(ok bool, provider, messageID, errMessage string) map[string]any {
	envelope := map[string]any{"ok": ok, "provider": provider}
	if messageID != "" {
		envelope["providerMessageId"] = messageID
	}
	if errMessage != "" {
		envelope["error"] = errMessage
	}
	return envelope
}

// emailMessage is a validated email.send input: bounded subject and bodies,
// at most 20 short metadata entries, and the resolved sender.
type emailMessage struct {
	to, from, subject, text, html string
	metadata                      map[string]string
}

// parseEmailMessage validates the tool input; a non-empty second value is
// the noop-envelope error message.
func parseEmailMessage(input map[string]any) (emailMessage, string) {
	msg := emailMessage{metadata: map[string]string{}}
	msg.to, _ = input["to"].(string)
	msg.subject, _ = input["subject"].(string)
	msg.text, _ = input["text"].(string)
	msg.html, _ = input["html"].(string)
	msg.from, _ = input["from"].(string)
	if msg.to == "" || msg.subject == "" || len(msg.subject) > emailSubjectMax {
		return msg, "email.send requires to and subject (subject ≤998 chars)"
	}
	if msg.text == "" && msg.html == "" {
		return msg, "email.send requires `text` or `html` (or both)."
	}
	if len(msg.text) > emailTextMax || len(msg.html) > emailHTMLMax {
		return msg, "email.send body exceeds the size cap"
	}
	if rawMetadata, ok := input["metadata"].(map[string]any); ok {
		if len(rawMetadata) > emailMetadataMax {
			return msg, "email.send metadata supports at most 20 entries"
		}
		for key, value := range rawMetadata {
			text, ok := value.(string)
			if !ok || key == "" || len(key) > 64 || len(text) > 256 {
				return msg, "email.send metadata entries must be short strings"
			}
			msg.metadata[key] = text
		}
	}
	return msg, ""
}

// emailRequest is one provider's wire shape for a message plus how it
// reports the message id (nil: the provider answers no id and one is
// synthesized).
type emailRequest struct {
	label     string
	url       string
	headers   map[string]string
	payload   map[string]any
	messageID func(body string) string
}

func jsonMessageID(body string) string {
	var parsed struct {
		ID string `json:"id"`
	}
	_ = json.Unmarshal([]byte(body), &parsed)
	return parsed.ID
}

func bearerJSONHeaders(envKey string) map[string]string {
	return map[string]string{
		"Authorization": "Bearer " + os.Getenv(envKey),
		"Content-Type":  "application/json",
	}
}

// emailProviderRequests builds the provider request for a validated message.
var emailProviderRequests = map[string]func(msg emailMessage) emailRequest{
	"resend": func(msg emailMessage) emailRequest {
		payload := map[string]any{"from": msg.from, "to": []string{msg.to}, "subject": msg.subject}
		if msg.text != "" {
			payload["text"] = msg.text
		}
		if msg.html != "" {
			payload["html"] = msg.html
		}
		if len(msg.metadata) > 0 {
			tags := make([]map[string]string, 0, len(msg.metadata))
			for name, value := range msg.metadata {
				tags = append(tags, map[string]string{"name": name, "value": value})
			}
			payload["tags"] = tags
		}
		return emailRequest{label: "Resend", url: "https://api.resend.com/emails",
			headers: bearerJSONHeaders("RESEND_API_KEY"), payload: payload, messageID: jsonMessageID}
	},
	"sendgrid": func(msg emailMessage) emailRequest {
		content := []map[string]string{}
		if msg.text != "" {
			content = append(content, map[string]string{"type": "text/plain", "value": msg.text})
		}
		if msg.html != "" {
			content = append(content, map[string]string{"type": "text/html", "value": msg.html})
		}
		payload := map[string]any{
			"personalizations": []map[string]any{{"to": []map[string]string{{"email": msg.to}}}},
			"from":             map[string]string{"email": msg.from},
			"subject":          msg.subject,
			"content":          content,
		}
		if len(msg.metadata) > 0 {
			payload["custom_args"] = msg.metadata
		}
		// SendGrid answers 202 + empty body; the id rides a response
		// header FetchHTTPTarget doesn't expose — synthesize best-effort.
		return emailRequest{label: "SendGrid", url: "https://api.sendgrid.com/v3/mail/send",
			headers: bearerJSONHeaders("SENDGRID_API_KEY"), payload: payload}
	},
	"simulator": func(msg emailMessage) emailRequest {
		payload := map[string]any{"to": msg.to, "from": msg.from, "subject": msg.subject}
		if msg.text != "" {
			payload["text"] = msg.text
		}
		if msg.html != "" {
			payload["html"] = msg.html
		}
		return emailRequest{label: "Simulator", url: simulatorEndpoint("/email/send"),
			headers: map[string]string{"content-type": "application/json"}, payload: payload, messageID: jsonMessageID}
	},
}

// executeEmailSend runs the email.send tool through the chokepoint deps.
// NEVER throws — every failure mode answers {ok:false, provider, error}.
func executeEmailSend(ctx context.Context, input map[string]any, deps *IntegrationDeps) map[string]any {
	start := time.Now()
	msg, errMessage := parseEmailMessage(input)
	if errMessage != "" {
		return emailEnvelope(false, "noop", "", errMessage)
	}

	settings := EmailSettings{}
	if deps != nil && deps.Email != nil {
		settings = deps.Email()
	}
	if msg.from == "" {
		msg.from = settings.From
	}
	if msg.from == "" {
		msg.from = os.Getenv("JANUSLY_MAILER_FROM")
	}
	if msg.from == "" {
		msg.from = "onboarding@resend.dev"
	}
	if !validEmailMailbox(msg.from) {
		return emailEnvelope(false, "noop", "", "configured mailer from address is invalid")
	}
	latency := func() int { return int(time.Since(start).Milliseconds()) }
	record := func(provider string, ok bool, errMessage string) {
		if deps != nil && deps.Record != nil {
			deps.Record("email.send", provider, ok, 0, errMessage, latency())
		}
	}

	// Per-org rate gate: over-limit converts to a clean envelope.
	if deps != nil && deps.RateLimit != nil {
		rateLimit := 100
		if deps.RateLimitPerMin != nil {
			rateLimit = deps.RateLimitPerMin("email", 100)
		}
		if errMessage := deps.RateLimit(ctx, "email.send", rateLimit); errMessage != "" {
			record("noop", false, errMessage)
			return emailEnvelope(false, "noop", "", errMessage)
		}
	}

	provider := resolveMailerProvider(settings.Provider)
	if deps == nil || deps.Post == nil {
		return emailEnvelope(false, "noop", "", "integration tools require run context")
	}
	buildRequest := emailProviderRequests[provider]
	if buildRequest == nil {
		message := "Mailer not configured. Set JANUSLY_MAILER_PROVIDER=resend|sendgrid and the matching API key."
		record("noop", false, message)
		return emailEnvelope(false, "noop", "", message)
	}
	request := buildRequest(msg)
	serialized, err := json.Marshal(request.payload)
	if err != nil {
		record(provider, false, "payload serialization failed")
		return emailEnvelope(false, provider, "", "payload serialization failed")
	}
	statusCode, body, errMessage := deps.Post(ctx, request.url, request.headers, serialized)
	if errMessage != "" {
		record(provider, false, errMessage)
		return emailEnvelope(false, provider, "", errMessage)
	}
	if statusCode < 200 || statusCode >= 300 {
		message := fmt.Sprintf("%s returned HTTP %d: %s", request.label, statusCode, truncateBody(body))
		record(provider, false, message)
		return emailEnvelope(false, provider, "", message)
	}
	messageID := fmt.Sprintf("%s-%d", provider, time.Now().UnixMilli())
	if request.messageID != nil {
		if messageID = request.messageID(body); messageID == "" {
			message := request.label + " response missing message id"
			record(provider, false, message)
			return emailEnvelope(false, provider, "", message)
		}
	}
	record(provider, true, "")
	return emailEnvelope(true, provider, messageID, "")
}
